## 1-CLASIFICACION CON NAIVE BAYES


### 1- IMPORTAMOS LOS DATOS

In [ ]:
import pandas as pd

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


In [ ]:
solicitantes = pd.read_excel('/content/drive/MyDrive/0. SolicitantesCrédito.xlsx')

In [ ]:
solicitantes

Unnamed: 0  Edad    Sexo Estado civil  Hijos  Perscargo  Estrato  \
0              1    38      NR       Casado      1          0        3   
1              2    51      NR     Separado      6          1        4   
2              3    37   Mujer       Casado      2          0        3   
3              4    29      NR     Separado      1          0        2   
4              5    42   Mujer       Casado      3          4        2   
...          ...   ...     ...          ...    ...        ...      ...   
8373        8374    31  Hombre      Soltero      2          0        5   
8374        8375    38  Hombre     Separado      0          0        3   
8375        8376    43   Mujer       Casado      1          2        2   
8376        8377    37   Mujer       Casado      1          2        3   
8377        8378    24      NR      Soltero      0          0        2   

     Nivel de estudios  Tipo de contrato  Ingresos  ...  Plazo    Cuota (COP)  \
0            Bachiller           Ninguno   5424576  ...     48   82691.506204   
1              Técnico           Ninguno   1144064  ...     36   22781.212598   
2            Tecnólogo        Indefinido   2069853  ...     36   46294.596244   
3            Tecnólogo           Ninguno   1893080  ...     24   69564.950015   
4      Especializacion           Ninguno   3000367  ...     24  112478.307883   
...                ...               ...       ...  ...    ...            ...   
8373         Bachiller        Indefinido   5891088  ...     48  115108.056436   
8374         Bachiller  Labor Contratada   3092079  ...     36   96789.562091   
8375         Tecnólogo           Ninguno   2542007  ...     36   49713.608249   
8376         Tecnólogo        Indefinido   5677067  ...     36  163602.818587   
8377         Tecnólogo        Indefinido   1950392  ...     36   69024.928055   

                ConoCliente          Garantia            RespPat  \
0        Buenas Referencias      Sin Codeudor     Sin Patrimonio   
1        Buenas Referencias      Sin Codeudor  Prop.sin Hipoteca   
2        Buenas Referencias  Con Codeudor(es)     Sin Patrimonio   
3        Buenas Referencias  Con Codeudor(es)     Sin Patrimonio   
4        Buenas Referencias  Con Codeudor(es)     Sin Patrimonio   
...                     ...               ...                ...   
8373  Buena Experiencia CFB      Sin Codeudor     Sin Patrimonio   
8374     Buenas Referencias      Sin Codeudor  Prop.sin Hipoteca   
8375     Buenas Referencias      Sin Codeudor     Sin Patrimonio   
8376     Buenas Referencias  Con Codeudor(es)     Sin Patrimonio   
8377     Buenas Referencias      Sin Codeudor     Sin Patrimonio   

                      MoraCome  PreApr  Prob.Default (PD)       Score  \
0     Con Experiencia sin Mora     1.0           0.041659  582.478943   
1     Con Experiencia sin Mora     1.0           0.234951  479.839966   
2     Con Experiencia sin Mora     0.0           0.170330  417.653107   
3     Con Experiencia sin Mora     1.0           0.201659  316.791107   
4     Con Experiencia sin Mora     1.0           0.140929  441.809570   
...                        ...     ...                ...         ...   
8373  Con Experiencia sin Mora     1.0           0.035120  659.471375   
8374  Con Experiencia sin Mora     1.0           0.101976  465.011536   
8375  Con Experiencia sin Mora     1.0           0.156707  423.237305   
8376  Con Experiencia sin Mora     1.0           0.037719  592.670837   
8377  Con Experiencia sin Mora     NaN                NaN         NaN   

           LGD  
0     0.063702  
1     0.465636  
2     0.290071  
3     0.278389  
4     0.203359  
...        ...  
8373  0.066962  
8374  0.165936  
8375  0.247763  
8376  0.051356  
8377       NaN  

[8378 rows x 35 columns]

In [ ]:

XD = solicitantes.dropna(subset=['PreApr'])

In [ ]:

XD = XD [['Edad','Ingresos','Egresos','Monto (EAD)']]

In [ ]:
XD

Edad  Ingresos  Egresos   Monto (EAD)
0       38   5424576  7585302  2.815030e+06
1       51   1144064  1640079  6.301439e+05
2       37   2069853  2831440  1.280540e+06
3       29   1893080  2482276  1.393414e+06
4       42   3000367  3630221  2.252986e+06
...    ...       ...      ...           ...
8371    48   4831392  3392105  3.366188e+06
8373    31   5891088  4292158  3.918572e+06
8374    38   3092079  3028098  2.677266e+06
8375    43   2542007  3513112  1.375112e+06
8376    37   5677067  5264273  4.525366e+06

[5842 rows x 4 columns]

In [ ]:

yd = solicitantes['PreApr'].dropna()

In [ ]:
yd

0       1.0
1       1.0
2       0.0
3       1.0
4       1.0
       ... 
8371    1.0
8373    1.0
8374    1.0
8375    1.0
8376    1.0
Name: PreApr, Length: 5842, dtype: float64

LIMPIAMOS DATOS


In [ ]:

vars_old = ['Edad','Ingresos','Egresos','Monto (EAD)']

vars_new = ['Edad','Hijos','Perscargo','Estrato','Ingresos','Egresos','Monto (EAD)','Plazo','Cuota (COP)']

df_new = solicitantes.dropna(subset=vars_new).copy()

df_old = solicitantes.dropna(subset=vars_old).copy()

print("Filas originales:", solicitantes.shape[0])
print("Filas limpias (modelo anterior, 4 vars):", df_old.shape[0])
print("Filas limpias (Reto 1, 9 vars):", df_new.shape[0])

XD_old = df_old[vars_old]
yd_old = df_old['PreApr']

XD_new = df_new[vars_new]
yd_new = df_new['PreApr']

XD = XD_new
yd = yd_new

Filas originales: 8378
Filas limpias (modelo anterior, 4 vars): 8378
Filas limpias (Reto 1, 9 vars): 8378


IMPLEMENTAMOS EL MODELO NAIVE BAYES


In [ ]:
### IMPLEMENTAMOS LAS LIBRERIAS NECESARIAS
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt


from sklearn.naive_bayes import GaussianNB

In [ ]:
from sklearn.metrics import confusion_matrix
from sklearn.naive_bayes import GaussianNB
import numpy as np
import pandas as pd

def metricas_binarias_desde_cm(cm):
    """Calcula métricas a partir de una matriz de confusión 2x2:
    [[VN, FP],
     [FN, VP]]
    """
    VN, FP, FN, VP = cm[0,0], cm[0,1], cm[1,0], cm[1,1]
    total = VN + FP + FN + VP

    exactitud = (VP + VN) / total if total else np.nan
    error = (FP + FN) / total if total else np.nan
    sensibilidad = VP / (VP + FN) if (VP + FN) else np.nan   # recall
    precision = VP / (VP + FP) if (VP + FP) else np.nan
    especificidad = VN / (VN + FP) if (VN + FP) else np.nan

    return {
        'VN': int(VN), 'FP': int(FP), 'FN': int(FN), 'VP': int(VP),
        'Exactitud': exactitud,
        'Error': error,
        'Precision': precision,
        'Sensibilidad': sensibilidad,
        'Especificidad': especificidad,
    }

def entrenar_y_evaluar(nombre, X, y):
    X_df = pd.DataFrame(X).copy()
    y_sr = pd.Series(y, name="y").copy()

    data = pd.concat([X_df, y_sr], axis=1)

    data = data.dropna()

    X_clean = data.iloc[:, :-1].to_numpy()
    y_clean = data.iloc[:, -1].to_numpy()

    y_clean = y_clean.astype(int)

    modelo = GaussianNB()
    modelo.fit(X_clean, y_clean)

    y_pred = modelo.predict(X_clean)  # mismo conjunto
    cm = confusion_matrix(y_clean, y_pred)

    metr = metricas_binarias_desde_cm(cm)
    metr['Modelo'] = nombre
    metr['Filas'] = len(y_clean)
    return modelo, y_pred, cm, metr

modelo_old, ypred_old, cm_old, metr_old = entrenar_y_evaluar('Anterior (4 vars)', XD_old, yd_old)
modelo_new, ypred_new, cm_new, metr_new = entrenar_y_evaluar('Reto 1 (9 vars)', XD_new, yd_new)


modeloNB = modelo_new
ydPred = ypred_new
ydArray = np.array(yd_new)
cm = cm_new


comparacion = pd.DataFrame([metr_old, metr_new]).set_index('Modelo')[
    ['Filas','VN','FP','FN','VP','Exactitud','Error','Precision','Sensibilidad','Especificidad']
]
comparacion


Filas    VN   FP    FN    VP  Exactitud     Error  \
Modelo                                                                 
Anterior (4 vars)   5842  2503  456   938  1945   0.761383  0.238617   
Reto 1 (9 vars)     5842  2539  420  1041  1842   0.749914  0.250086   

                   Precision  Sensibilidad  Especificidad  
Modelo                                                     
Anterior (4 vars)   0.810079      0.674644       0.845894  
Reto 1 (9 vars)     0.814324      0.638918       0.858060

MATRIZ DE CONFUSION

In [ ]:
# Matrices de confusión (formato: filas = y real, columnas = y pred)
import pandas as pd

def cm_df(cm):
    return pd.DataFrame(cm, index=['Real 0','Real 1'], columns=['Pred 0','Pred 1'])

print("Matriz de confusión — Modelo anterior (4 vars):")
display(cm_df(cm_old))
print("\nMatriz de confusión — Modelo Reto 1 (9 vars):")
display(cm_df(cm_new))


print("\nMétricas — Modelo anterior (4 vars):")
for k in ['Exactitud','Error','Precision','Sensibilidad','Especificidad']:
    print(f"{k}: {comparacion.loc['Anterior (4 vars)', k]:.4f}")

print("\nMétricas — Modelo Reto 1 (9 vars):")
for k in ['Exactitud','Error','Precision','Sensibilidad','Especificidad']:
    print(f"{k}: {comparacion.loc['Reto 1 (9 vars)', k]:.4f}")

Matriz de confusión — Modelo anterior (4 vars):


Pred 0  Pred 1
Real 0    2503     456
Real 1     938    1945


Matriz de confusión — Modelo Reto 1 (9 vars):


Pred 0  Pred 1
Real 0    2539     420
Real 1    1041    1842


Métricas — Modelo anterior (4 vars):
Exactitud: 0.7614
Error: 0.2386
Precision: 0.8101
Sensibilidad: 0.6746
Especificidad: 0.8459

Métricas — Modelo Reto 1 (9 vars):
Exactitud: 0.7499
Error: 0.2501
Precision: 0.8143
Sensibilidad: 0.6389
Especificidad: 0.8581


METRICAS DE EVALUACION DEL MODELO

Generalización del modelo

In [ ]:

pendientes = solicitantes[solicitantes['PreApr'].isna()].copy()
print("Solicitantes sin PreApr (antes de limpiar NaN en X):", pendientes.shape[0])


pendientes_limpio = pendientes.dropna(subset=vars_new).copy()
print("Solicitantes sin PreApr (listos para predecir):", pendientes_limpio.shape[0])

if pendientes_limpio.shape[0] > 0:
    X_gener = pendientes_limpio[vars_new]
    y_gener = modeloNB.predict(np.array(X_gener))
    pendientes_limpio['PreApr_Pred'] = y_gener
    pendientes_limpio[['PreApr_Pred'] + vars_new].head()
else:
    print("No hay solicitantes completos (sin NaN en X) para generar predicciones.")

Solicitantes sin PreApr (antes de limpiar NaN en X): 2536
Solicitantes sin PreApr (listos para predecir): 2536


En este ejercicio se re-entrenó un modelo de Gaussian Naive Bayes para la
predicción de la variable PreApr, ampliando el conjunto de variables
independientes de 4 a 9, incluyendo variables demográficas y financieras:
Edad, Hijos, Personas a cargo, Estrato, Ingresos, Egresos, Monto (EAD),
Plazo y Cuota (COP).


Los resultados obtenidos muestran que el modelo con un mayor número de
variables no presenta una mejora en el desempeño global frente al modelo
anterior. En particular, la exactitud disminuyó de 0.7614 a 0.7499 y la
sensibilidad se redujo de 0.6746 a 0.6389, indicando una menor capacidad
del modelo para identificar correctamente las solicitudes aprobadas.
Aunque se observa una leve mejora en la precisión y la especificidad,
estos incrementos no compensan la pérdida en capacidad de predicción.

Esto sugiere que agregar nuevas variables
financieras introduce correlaciones que afectan el supuesto de independencia condicional propio del modelo Gaussian Naive
Bayes. Por eso se puede concluir que un modelo más
simple con menos variables resulta más efectivo para este caso.